lib/jsverify.d.ts

```diff
@@ -78,11 +93,12 @@ declare namespace JSVerify {
 
   //Combinators
   function nonShrink<T>(arb: Arbitrary<T>): Arbitrary<T>;
-  function either<T, U>(arbA: Arbitrary<T>, arbB: Arbitrary<U>): Arbitrary<T | U>;
+  function either<T, U>(arbA: Arbitrary<T>, arbB: Arbitrary<U>): Arbitrary<Either<T, U>>;
   function pair<T, U>(arbA: Arbitrary<T>, arbB: Arbitrary<U>): Arbitrary<[T, U]>;
 
   function tuple(arbs: Arbitrary<any>[]): Arbitrary<any[]>;
-  function sum(arbs: Arbitrary<any>[]): Arbitrary<any>;
+  function sum<T>(arbs: Arbitrary<T>[]): Arbitrary<Addend<T>>;
+  function sum(arbs: Arbitrary<any>[]): Arbitrary<Addend<any>>;
```

Comment:
I think jvanbruegge is right. Getting this right is tricky. We could do something like provide `n` overloads with parameters being tuple types, then a fallback to the any version if the size of the array exceeds `n`. This is what most typescript definitions for `Promise.all` do.

On a side note, [Variadic kinds](https://github.com/Microsoft/TypeScript/issues/5453) would probably help here.